auth.py

```
@@ -105,4 +105,100 @@ def complex_auth_function(user, pass_word, email, phone, address, city, state, z
+def calculate_user_score(total_points, completed_tasks):
+    return total_points / completed_tasks
+
+# Issue 19: Performance - Inefficient string concatenation in loop
+def build_user_report(users):
+    report = ""
+    for user in users:
+        report = report + f"User: {user['name']}, Email: {user['email']}\n"
+    return report
+
+# Issue 20: Best Practices - Using mutable default argument
+def add_user_permissions(user_id, permissions=[]):
```

Ruff (0.12.2)

Do not use mutable data structures for argument defaults

Replace with `None`; initialize within function

(B006)